presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java

```diff
@@ -0,0 +1,244 @@
+                probeNode = ((AggregationNode) probeNode).getSource();
+            }
+            else if (probeNode instanceof FilterNode) {
+                probeNode = ((FilterNode) probeNode).getSource();
+            }
+            else {
+                break;
+            }
+        }
+        log.warn("createDynamicFilter: unsupported node %s", probeNode);
+        return Optional.empty();
+    }
+
+    public static Optional<DynamicFilter> create(JoinNode node, PlanNode probeNode, List<Symbol> probeSymbols, PlanNode buildNode, List<Symbol> buildSymbols, TypeProvider types)
+    {
+        if (node.getType() != JoinNode.Type.INNER) {
+            log.debug("createDynamicFilter: dynamic filters are supported only for inner-joins");
+            return Optional.empty();
+        }
+        if (!node.getDistributionType().equals(Optional.of(JoinNode.DistributionType.REPLICATED))) {
+            log.debug("createDynamicFilter: dynamic filters are supported only for replicated joins");
+            return Optional.empty();
+        }
+
+        int numOfJoinSymbols = probeSymbols.size();
+        if (numOfJoinSymbols != buildSymbols.size()) {
+            // TODO: consider matching some of the symbols for partial dynamic filtering
+            log.debug("createDynamicFilter: buildSymbols' size (%d) doesn't match probeSymbols' size (%d)",
+                      numOfJoinSymbols, probeSymbols.size());
+            return Optional.empty();
+        }
+
+        return findTableScanNode(probeNode, probeSymbols).map(scanNode -> {
+            ImmutableList.Builder<DynamicFilter.Descriptor> dynamicFilters = new ImmutableList.Builder<>();
+            for (int i = 0; i < numOfJoinSymbols; ++i) {
+                Symbol probeSymbol = probeSymbols.get(i);
+                Symbol buildSymbol = buildSymbols.get(i);
+
+                Type buildType = types.get(buildSymbol);
+                int buildChannel = buildNode.getOutputSymbols().indexOf(buildSymbol);
+
+                ColumnHandle probeColumn = scanNode.getAssignments().get(probeSymbol);
+                dynamicFilters.add(new DynamicFilter.Descriptor(buildType, buildChannel, probeColumn));
+            }
+            log.debug("createDynamicFilter: dynamic filters initialized for join node: %s", node);
+            return new DynamicFilter(dynamicFilters.build());
+        });
+    }
+
+    public class Collector
+    {
+        private final ImmutableList.Builder<Page> pageBuilder;
+        private final PageChannelSelector selector;
+
+        public Collector()
+        {
+            this.pageBuilder = new ImmutableList.Builder<>();
+            int[] buildChannels = descriptors.stream().mapToInt(d -> d.buildChannel).toArray();
+            this.selector = new PageChannelSelector(buildChannels);
+        }
+
+        public void add(Page page)
+        {
+            // Collect only the columns which are relevant for the JOIN.
+            pageBuilder.add(selector.apply(page));
```

Comment:
once the remaining rows is exceeded, you could drop references to collected pages

Reply:
Done at 39a2eb47da.